core/core.go

```diff
@@ -167,70 +141,69 @@ func (c *core) listen() {
 				return
 			}
 			newSignal.SetBytes("response", response)
+
 			responder <- newSignal
 		}(newSignal)
 	}
 }
 
-func (c *core) SetState(key string, state spec.State) {
-	c.Mutex.Lock()
-	defer c.Mutex.Unlock()
-	c.States[key] = state
+func (c *core) listenToSignal() {
+	c.Log.V(11).Debugf("call Core.listenToSignal")
+
+	listener := make(chan os.Signal, 1)
+	signal.Notify(listener, os.Interrupt, os.Kill)
+
+	<-listener
+
+	c.Shutdown()
 }
 
 func (c *core) Shutdown() {
 	c.Log.V(12).Debugf("call Core.Shutdown")
 
-	// TODO close gateway
-	// TODO stop listening
-	// TODO wait for impulses to be processed
-	// TODO backup state
-}
+	c.TextGateway.Close()
 
-func (c *core) Trigger(imp spec.Impulse) (spec.Impulse, error) {
-	c.Log.V(12).Debugf("call Core.Trigger")
+	for {
+		impulsesInProgress := atomic.LoadInt64(&c.ImpulsesInProgress)
+		if impulsesInProgress == 0 {
+			// As soon as all impulses are processed we can go ahead to shutdown the
+			// core.
+			break
+		}
 
-	// Track state.
-	impState, err := imp.GetState(common.DefaultStateKey)
-	if err != nil {
-		return nil, maskAny(err)
+		time.Sleep(100 * time.Millisecond)
 	}
-	impState.SetCore(c)
-	coreState, err := c.GetState(common.DefaultStateKey)
+
+	bytes, err := json.Marshal(c)
 	if err != nil {
-		return nil, maskAny(err)
+		c.Log.V(3).Errorf("%#v", maskAny(err))
 	}
-	coreState.SetImpulse(imp)
+
+	fmt.Printf("\n")
+	fmt.Printf("%s\n", bytes)
+	fmt.Printf("\n")
+
+	os.Exit(0)
```

Comment:
The state still needs to be back upped properly. Some restore mechanism also needs to be provided.